### ISSUE TITLE
Express version info dependency as dependency

### ISSUE THREAD

**Issue by PDoakORNL** [QMCPACK/qmcpack#689]
So I've done it like this
```cmake
execute_process(
  COMMAND git rev-parse --abbrev-ref HEAD
  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
  OUTPUT_VARIABLE GIT_BRANCH
  OUTPUT_STRIP_TRAILING_WHITESPACE
)

# Get the latest abbreviated commit hash of the working branch
execute_process(
  COMMAND git log -1 --format=%h
  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
  OUTPUT_VARIABLE GIT_COMMIT_HASH
  OUTPUT_STRIP_TRAILING_WHITESPACE
)

configure_file(
  ${CMAKE_SOURCE_DIR}/include/version.h.in
  ${CMAKE_BINARY_DIR}/generated/version.h
)

include_directories(${CMAKE_BINARY_DIR}/generated)
```

**Comment by markdewing** [comment]
One drawback with this approach is that it only runs when cmake is run.

**Comment by PDoakORNL** [comment]
I guess I would expect a build used in production or CI to be started clean and not on top of a previous build. Maybe that's naive. I don't see how the existing one would trigger if cmake didn't run either, if it does I'd like to understand that.

**Comment by prckent** [comment]
Since the above version "will work" the question returns to whether the current implementation can be made reliable. I would rather use something reliable that requires CMake to be rerun to be fully up to date than use something unreliable.

**Comment by markdewing** [comment]
The issue is that src/qmc_common.cpp and src/QMCApp/QMCMain.cpp both depend on qmcpack_version.h (and so git-rev-tmp.h), but only the dependency on src/qmc_common.cpp is in the CMakeLists.txt.

**Comment by ye-luo** [comment]
@PDoakORNL When developing codes, we start more often from current build instead of from scratch. That is why update commit id should be per make not per cmake.
@markdewing Can we put the revision printing in src/qmc_common.cpp as a function? Will this avoid dependency issue? It can be also useful for other tools to print revision.

**Comment by markdewing** [comment]
@ye-luo Yes, putting the revision printing in one spot should centralize the dependency, at least.

**Comment by PDoakORNL** [comment]
Ok I get it now.  If this is turned into an add_custom_target then the execution of the command can be made a Makefile level dependency by using add_dependency on targets and this parallel make race-condition should stop. That should be just as dependable as my way but retain the functionality you all want.

**Comment by prckent** [comment]
Closed by #725
